examples/test-exec-doc.yml

```diff
@@ -0,0 +1,221 @@
+# Name of the workflow
+name: Run Tests on Push to Main
+
+# Set permissions required for the workflow
+permissions:
+  contents: write           # Allows writing repository contents
+  pull-requests: write      # Allows creating and modifying pull requests
+  id-token: write           # Allows generating tokens for authentication
+  issues: write             # Allows creating and modifying issues
+
+# Define the events that trigger the workflow
+on:
+  push:
+    paths:
+      - '**/*.md'           # Trigger on push to any Markdown file
+  workflow_dispatch:        # Allows manual triggering of the workflow
+
+jobs:
+  run-tests:
+    runs-on: ubuntu-latest  # Run the job on the latest Ubuntu runner
+
+    steps:
+    - name: Checkout repository
+      uses: actions/checkout@v2
+      with:
+        fetch-depth: 0        # Fetch all history for accurate diff
+
+    - name: Check for changed Markdown files
+      id: check_changes
+      run: |
+        # Check if this is the initial commit (all zeros SHA)
+        if [ "${{ github.event.before }}" = "0000000000000000000000000000000000000000" ]; then
+          echo "Initial commit detected."
+
+          # Determine the default branch (e.g., main)
+          if git show-ref --verify --quiet refs/remotes/origin/main; then
+            DEFAULT_BRANCH="main"
+          elif git show-ref --verify --quiet refs/remotes/origin/master; then
+            DEFAULT_BRANCH="master"
+          else
+            echo "Default branch is not 'main' or 'master'. Getting all Markdown files."
+            CHANGED_MARKDOWN_FILES=$(git ls-files '*.md')
+            echo "Markdown files to test:"
+            echo "$CHANGED_MARKDOWN_FILES"
+            echo "changed=true" >> $GITHUB_OUTPUT
+            echo "changed_files<<EOF" >> $GITHUB_OUTPUT
+            echo "$CHANGED_MARKDOWN_FILES" >> $GITHUB_OUTPUT
+            echo "EOF" >> $GITHUB_OUTPUT
+            exit 0
+          fi
+          echo "Default branch is $DEFAULT_BRANCH"
+          
+          # Get the last commit SHA of the default branch
+          UPSTREAM_SHA=$(git rev-parse origin/$DEFAULT_BRANCH)
+          echo "Upstream SHA: $UPSTREAM_SHA"
+          
+          # List changed Markdown files by comparing with the upstream branch
+          CHANGED_MARKDOWN_FILES=$(git diff --name-only $UPSTREAM_SHA ${{ github.sha }} | grep -E '\.md$' || true)
```

Comment:
Why is this logic only needed where this is the first commit, but not needed in cases where this is the nth commit? Isn't it possible for people to use branches other than main in pre-existing repositories?

Reply:
it is needed for the nth commit and is handled by the else block below, where the nth commit and n-1th commit are compared to see the changes and only track those files. for the first commit, it tracks changes based on the last commit of the upstream branch

Reply:
I was more so referring to the branch checks where you check if the default branch is main or master, but I accidentally highlighted the entire section (Line 36-50). It didn't make sense to me as to why you only check what their default branch is on the first commit when they could have a different default branch on any commit.

Reply:
because in those cases, with the nth commit where n>1, there would be an n-1th commit in that very branch. The only time you need to check the default branch is when its the first commit of a new branch off of that branch. Otherwise, that new branch has the previous commit